### DISCUSSION TITLE
Issues with and suggestions for C# error logging (Print, PushWarning, PushError)

### DISCUSSION THREAD

**Discussion by zorbathut** [godotengine/godot-proposals#5647] [Scripting]
This is going to be a kind of rambly discussion, because there's a few issues that interlock.

I've got a C# program that I want Info, Warning, and Error log channel outputs for. At the moment I'm using GD.Print(), GD.PushWarning(), and GD.PushError(). This has some problems.

First, because GD.Print() goes to the Output tab and GD.Push*() go to the Debugger tab, this means the output isn't interlaced. This makes it very hard to debug warnings and errors with info output; practically speaking, this means that when I'm debugging I don't have a Warning channel, because I just print everything through Warning.

Second, Output-tab items don't come with a callstack. Sometimes I *really* want to know where a log line came from; there's just no context provided.

Third, everything under the Errors tab has a prefix of the last entry in the callstack. This is reasonably OK for Godot-engine stuff, but it fails miserably with GD.Push*(). A screenshot:

![image](https://user-images.githubusercontent.com/160754/197453419-250f3f7f-e371-4ce6-a002-8a77ec69d1f1.png)

The part I actually care about here is "Attempted construction with non-component type . . .", but that's hidden waaaay off on the right - the only way I can even see it is by mouseovering the log line (as I'm doing here), because if I click the dropdown, it doesn't wordwrap!

I've fixed this locally by just removing the callstack prefix:

![image](https://user-images.githubusercontent.com/160754/197453525-a4619525-d3ab-4058-bb38-48ddb5afed6b.png)

but this obviously removes that prefix from other errors as well.

I'm not sure if there's anyone else with my requirements. But I do have to say that Unity did a pretty good job here; having all log entries in the same list, with click-for-callstack on all of them, and prioritizing the user output over boilerplate content, really does seem like the right way to go. I can work with my changes but other people might be having similar issues.

Alternatively, I'm pretty sure Unity's log system is *not* the paragon of perfection, and we could probably make something better. But IMO it should be at least as capable as Unity's :)

I'd be happy to work on implementing agreed-upon changes here, and if there's no agreement, I'll probably implement my own changes and then just sit on them. But I'd prefer if I could contribute something :)

**Comment by zorbathut** [comment]
For posterity, here's the callstack-prefix-removal diff:

![image](https://user-images.githubusercontent.com/160754/197453910-4da67fab-36aa-4d77-a9bf-579384cba75d.png)

Obviously I'm kind of hesitant to send a pull request that does nothing but remove functionality :V

**Comment by Calinou** [comment]
See https://github.com/godotengine/godot/issues/67451.
